mitxgraders/helpers/calc.py

```diff
@@ -446,10 +447,18 @@ def parse_algebra(self):
         # Predefine recursive variables.
         expr = Forward()
 
-        # Handle variables passed in. They must start with letters/underscores
+        # Handle variables passed in. They must start with a letter
         # and may contain numbers afterward.
-        inner_varname = Combine(Word(alphas + "_", alphanums + "_") + ZeroOrMore("'"))
-        varname = Group(inner_varname)("variable")
+        inner_varname = Combine(Word(alphas, alphanums + "_") + ZeroOrMore("'"))
+        # Alternative variable name in tensor format
+        # Tensor name must be alphas only, indices may be alphanumeric
+        # e.g., U_{ijk}^{123}
+        indices = Literal("{") + Word(alphanums) + Literal("}")
+        tensor_lower = Combine(Word(alphas) + Literal("_") + indices)
+        tensor_upper = Combine(Word(alphas) + Literal("^") + indices)
+        tensor_mixed = Combine(Word(alphas) + Literal("_") + indices + Literal("^") + indices)
+        # Try for tensor mixed first, then upper, then lower, then generic variable
+        varname = Group(tensor_mixed | tensor_upper | tensor_lower | inner_varname)("variable")
```

Comment:
Also, in an ideal world, I think we would allow `T_i^{123}` and `T_{i}^{123}` **and interpret them the same way**.

(To be clear, if we can't interpret them the same, then I think disallowing `T_i^{123}` is the best option.)

This could be a future enhancement if it is desired.

Reply:
I think I want to keep braces to indicate indices for tensors. If we ever make a special sampler for tensors (eg, specifying traces, symmetries, etc), we're going to want tensors to be very specific. Equally so if we ever do abstract index notation for relativity.